Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm

```diff
@@ -307,12 +311,32 @@
             if (!tab)
                 break;
 
-            context->handleContentRuleListNotificationForTab(*tab, url, result.second);
+            savedMatchedRule |= context->handleContentRuleListNotificationForTab(*tab, url, result.second);
+
             break;
         }
     }
+
+    if (!savedMatchedRule)
+        return;
+
+    if (!m_purgeOldMatchedRulesTimer) {
+        m_purgeOldMatchedRulesTimer = makeUnique<WebCore::Timer>(*this, &WebExtensionController::purgeOldMatchedRules);
+        m_purgeOldMatchedRulesTimer->start(Seconds::fromMinutes(purgeMatchedRulesIntervalInMinutes), Seconds::fromMinutes(purgeMatchedRulesIntervalInMinutes));
+    }
 }
 
+void WebExtensionController::purgeOldMatchedRules()
+{
+    WallTime earliestDateToKeep = WallTime::now() - Seconds::fromMinutes(purgeMatchedRulesIntervalInMinutes);
+
+    bool stillHaveRules = false;
+    for (auto& extensionContext : m_extensionContexts)
+        stillHaveRules |= extensionContext->purgeMatchedRulesFromBefore(earliestDateToKeep);
+
+    if (!stillHaveRules)
+        m_purgeOldMatchedRulesTimer = nullptr;
```

Comment:
I don't think you need the `bool` return and `stillHaveRules`. This can be:

```
if (m_matchedRules.isEmpty())
    m_purgeOldMatchedRulesTimer = nullptr;
```

Reply:
This is in the controller so it applies across all extensions - `m_matchedRules` is per extension.

Reply:
So that means we need to keep track of if _any_ extensions still have rules - not just one of them.